### ISSUE TITLE
Failure to read flash

### ISSUE THREAD

**Issue by shenki** [open-power/skiboot#158]
This system has a problem where it fails to load anything but ff's from flash. While there's not much that skiboot could do, perhaps we could detect this and print a helpful message before aborting?

```
[   37.488023096,5] OPAL v5.10-3-g90e34127e356 starting...
[   37.488029257,7] initial console log level: memory 7, driver 5
[   37.488032834,6] CPU: P8 generation processor (max 8 threads/core)
[   37.488036104,7] CPU: Boot CPU PIR is 0x0028 PVR is 0x004d0200
[   37.488039621,7] CPU: Initial max PIR set to 0x1fff
[   37.488457398,7] OPAL table: 0x300f9630 .. 0x300f9b80, branch table: 0x30002000
[   37.488463353,7] Assigning physical memory map table for unused
[   37.488467797,7] FDT: Parsing fdt @0xff00000
[   37.489211974,3] DT: dt_attach_root failed, duplicate sensor@33
[   37.489216814,3] DT: Found duplicate node: sensor@33
[   37.490234335,6] CHIP: Initialised chip 0 from xscom@3fc0000000000
[   37.490283983,5] CHIP: Chip ID 0000 type: P8 DD2.0
[   37.490287533,7] XSCOM: Base address: 0x3fc0000000000
[   37.490296146,7] XSTOP: XSCOM addr = 0x2010c82, FIR bit = 31
[   37.490299974,6] MFSI 0:0: Initialized
[   37.490302530,6] MFSI 0:2: Initialized
[   37.490305064,6] MFSI 0:1: Initialized
[   37.490497865,5] LPC: LPC[000]: Initialized, access via XSCOM @0xb0020
[   37.490505900,7] LPC: Default bus on chip 0x0
[   37.490572317,6] MEM: parsing reserved memory from node /ibm,hostboot/reserved-memory
[   37.490589713,7] HOMER: Init chip 0
[   37.490592803,7]   PBA BAR0 : 0x0000000ffd800000
[   37.490596008,7]   PBA MASK0: 0x0000000000300000
[   37.490599084,7]   HOMER Image at 0xffd800000 size 4MB
[   37.490603194,7]   PBA BAR2 : 0x4000000ffda00000
[   37.490606386,7]   PBA MASK2: 0x0000000000000000
[   37.490609386,7]   SLW Image at 0xffda00000 size 1MB
[   37.490613225,7]   PBA BAR3 : 0x0000000fff800000
[   37.490616315,7]   PBA MASK3: 0x0000000000700000
[   37.490619297,7]   OCC Common Area at 0xfff800000 size 8MB
[   37.490761831,5] PLAT: Fixing up PSI BAR on chip 0 BAR=3fffe80000001
[   37.490792125,7] AST: PNOR LPC offset: 0x0e000000
[   37.490864687,5] PLAT: Using virtual UART
[   37.491135392,7] UART: Using LPC IRQ 4
[   37.493632053,5] PLAT: Detected Palmetto platform
[   37.493709655,5] PLAT: Detected BMC platform AMI
[   37.494248008,5] CENTAUR: Found centaur for chip 0x0 channel 1
[   37.494337047,5] CENTAUR:   FSI host: 0x0 cMFSI0 port 2
[   37.494604771,5] CPU: All 32 processors called in...
[   39.164168133,3] FFS: Reading the flash has returned all 0xFF.
[   39.164307097,3]      Are you reading erased flash?
[   39.164409433,3]      Is something else using the flash controller?
[   39.164512657,4] FLASH: No ffs info; using raw device only
[   39.164609471,5] BT: Interface initialized, IO 0x00e4
[   40.113582618,4] NVRAM: Query before is done loading
[   40.113663743,4] NVRAM: Waiting for load
[   40.113736824,2] NVRAM: Failed to load
[   40.113792399,4] NVRAM: Query before is done loading
[   40.113889356,4] NVRAM: Waiting for load
[   40.113953471,2] NVRAM: Failed to load
[   40.114010015,5] STB: Found ibm,secureboot-v1
[   40.114073771,4] NVRAM: Query before is done loading
[   40.114194206,4] NVRAM: Waiting for load
[   40.114251631,2] NVRAM: Failed to load
[   40.114314655,5] STB: secure mode off
[   40.115223874,4] NVRAM: Query before is done loading
[   40.115319247,4] NVRAM: Waiting for load
[   40.115368959,2] NVRAM: Failed to load
[   40.115432158,5] STB: trusted mode off
[   40.115487970,5] OPAL: Using OPAL UART console
[   40.115545914,4] NVRAM: Query before is done loading
[   40.115609106,4] NVRAM: Waiting for load
[   40.115668828,2] NVRAM: Failed to load
[   40.115800364,5] SLW: HB-provided idle states property found
[   40.115891540,5] SLW: Enabling: nap
[   40.115943668,5] SLW: Enabling: fastsleep_
[   40.116004269,5] SLW: Enabling: winkle
[   40.116062178,4] NVRAM: Query before is done loading
[   40.116140425,4] NVRAM: Waiting for load
[   40.116197455,2] NVRAM: Failed to load
[   40.116270855,4] FLASH: Can't load resource id:2. No system flash found
[   40.116272273,7] FLASH: Queueing preload of 1/0
[   40.116296696,6] NX0: RNG BAR set to 0x0003ffff40000000..0x0003ffff40000fff
[   40.116301583,6] NX0: Crypto at 0x2010000
[   40.116303740,7] NX0:   DMA 0x00000188eb80002c
[   40.116305272,6] NX0:   SYM CT already set to 1
[   40.116306319,6] NX0:   SYM CI already set to 1
[   40.116308181,7] NX0:   SYM Config 0x8008010000000000
[   40.116310432,6] NX0:   ASYM CT already set to 2
[   40.116311468,6] NX0:   ASYM CI already set to 1
[   40.116312990,7] NX0:   ASYM Config 0x8020020000000000
[   40.116315331,7] NX0:   CRB Input Queue 0x0000000000000000
[   40.116317071,7] NX0:   Engine Enable 0x8000000000000003
[   40.116318461,6] NX0: Crypto Coprocessors Disabled (not supported)
[   40.116320093,6] NX0: 842 at 0x2010000
[   40.116323191,7] NX0:   DMA 0x00000188ff8000af
[   40.116325232,6] NX0:   842 CT set to 3
[   40.116326185,6] NX0:   842 CI already set to 1
[   40.116327615,7] NX0:   842 Config 0x00080300f8000001
[   40.116330415,7] NX0:   Engine Enable 0x8000000000000003
[   40.116331810,6] NX0: 842 Coprocessor Enabled
[   40.116397988,5] Chip 0 Found PBCQ0 at /xscom@3fc0000000000/pbcq@2012000
[   40.116402108,7] PHB3[0:0]: X[PE]=0x02012000 X[PCI]=0x09012000 X[SPCI]=0x09013c00
[   40.116405885,7] PHB3[0:0] REGS     = 0x0003fffe40000000 [4k]
[   40.116408127,7] PHB3[0:0] PCIBAR   = 0x0003fffe40000000
[   40.116410516,7] PHB3[0:0] MMIO0    = 0x0000200000000000 [0x0000010000000000]
[   40.116412766,7] PHB3[0:0] MMIO1    = 0x00003fe000000000 [0x0000000080000000]
[   40.116414895,7] PHB3[0:0] BAREN    = 0xf800000000000000
[   40.116417647,7] PHB3[0:0] NEWBAREN = 0xf800000000000000
[   40.116419493,7] PHB3[0:0] IRSNC    = 0x0100000000000000
[   40.116421165,7] PHB3[0:0] IRSNM    = 0xff00000000000000
[   40.116422591,7] PHB3[0:0] LSI      = 0xff00000000000000
[   40.116429347,5] Chip 0 Found PBCQ1 at /xscom@3fc0000000000/pbcq@2012400
[   40.116433352,7] PHB3[0:1]: X[PE]=0x02012400 X[PCI]=0x09012400 X[SPCI]=0x09013c40
[   40.116436438,7] PHB3[0:1] REGS     = 0x0003fffe40100000 [4k]
[   40.116439914,7] PHB3[0:1] PCIBAR   = 0x0003fffe40100000
[   40.116442378,7] PHB3[0:1] MMIO0    = 0x0000210000000000 [0x0000010000000000]
[   40.116445247,7] PHB3[0:1] MMIO1    = 0x00003fe080000000 [0x0000000080000000]
[   40.116447439,7] PHB3[0:1] BAREN    = 0xf800000000000000
[   40.116450434,7] PHB3[0:1] NEWBAREN = 0xf800000000000000
[   40.116452492,7] PHB3[0:1] IRSNC    = 0x0200000000000000
[   40.116454364,7] PHB3[0:1] IRSNM    = 0xff00000000000000
[   40.116455759,7] PHB3[0:1] LSI      = 0xff00000000000000
[   40.116462693,5] Chip 0 Found PBCQ2 at /xscom@3fc0000000000/pbcq@2012800
[   40.116466666,7] PHB3[0:2]: X[PE]=0x02012800 X[PCI]=0x09012800 X[SPCI]=0x09013c80
[   40.116469238,7] PHB3[0:2] REGS     = 0x0003fffe40200000 [4k]
[   40.116471754,7] PHB3[0:2] PCIBAR   = 0x0003fffe40200000
[   40.116473993,7] PHB3[0:2] MMIO0    = 0x0000220000000000 [0x0000010000000000]
[   40.116476714,7] PHB3[0:2] MMIO1    = 0x00003fe100000000 [0x0000000080000000]
[   40.116478820,7] PHB3[0:2] BAREN    = 0xf800000000000000
[   40.116481927,7] PHB3[0:2] NEWBAREN = 0xf800000000000000
[   40.116484189,7] PHB3[0:2] IRSNC    = 0x0300000000000000
[   40.116486453,7] PHB3[0:2] IRSNM    = 0xff00000000000000
[   40.116487855,7] PHB3[0:2] LSI      = 0xff00000000000000
[   40.116520567,7] PHB#0000:00:00.0 PCI: Registered PHB
[   40.116522963,6] PHB#0000: Found /pciex@3fffe40000000 @[0:0]
[   40.116525037,6] PHB#0000:   M32 [0x00003fe000000000..0x00003fe07fffffff]
[   40.116526801,6] PHB#0000:   M64 [0x0000200000000000..0x000020ffffffffff]
[   40.116528794,7] PHB#0000: Base location code not found !
[   40.116530127,6] PHB#0000: Max link speed: GEN3
[   40.116531317,7] PHB#0000: Override lane equalization settings:
[   40.116532525,7] PHB#0000:   0x3333333333333333 0x3333333333333333
[   40.116534255,7] PHB#0000:   0x3333333333333333 0x3333333333333333
[   40.116714624,7] IRQ: Registering 0800..0ff7 ops @0x300c0e78 (data 0x3052b5f8)
[   40.116717064,7] IRQ: Registering 0ff8..0fff ops @0x300c0ea8 (data 0x3052b5f8)
[   40.116719007,7] PHB#0000: Initializing PHB...
[   40.120616548,4] FLASH: Can't load resource id:0. No system flash found
[   40.120694942,4] FLASH: Can't load resource id:1. No system flash found
[   40.176471690,3] CAPP: Error loading ucode lid. index=200ea
[   40.296526803,4] NVRAM: Query before is done loading
[   40.296646955,4] NVRAM: Waiting for load
[   40.296721835,2] NVRAM: Failed to load
[   40.296776989,4] NVRAM: Query before is done loading
[   40.296850193,4] NVRAM: Waiting for load
[   40.296932302,2] NVRAM: Failed to load
[   40.296995281,4] NVRAM: Query before is done loading
[   40.297081640,4] NVRAM: Waiting for load
[   40.297140311,2] NVRAM: Failed to load
[   40.297188089,4] NVRAM: Query before is done loading
[   40.297250960,4] NVRAM: Waiting for load
[   40.297301096,2] NVRAM: Failed to load
[   40.297441595,4] NVRAM: Query before is done loading
[   40.297544267,4] NVRAM: Waiting for load
[   40.297601960,2] NVRAM: Failed to load
[   40.307934851,5] PCI: Resetting PHBs and training links...
[   41.410518629,5] PCI: Probing slots...
[   43.057372728,5] PHB#0000:00:00.0 [ROOT] 1014 03dc R:00 C:060400 B:01..ff SLOT=Slot2 
[   43.057588179,5] PHB#0001:00:00.0 [ROOT] 1014 03dc R:00 C:060400 B:01..07 SLOT=Backplane PLX 
[   43.057720610,5] PHB#0001:01:00.0 [SWUP] 10b5 8718 R:aa C:060400 B:02..07 LOC_CODE=Backplane PLX
[   43.057869165,5] PHB#0001:02:01.0 [SWDN] 10b5 8718 R:aa C:060400 B:03..04 
[   43.121705167,5] PHB#0001:03:00.0 [ETOX] ffff ffff R:ff C:ffffff B:04..04 LOC_CODE=Backplane PLX
[   43.121938881,5] PHB#0001:02:02.0 [SWDN] 10b5 8718 R:aa C:060400 B:05..05 
[   43.122065819,5] PHB#0001:05:00.0 [EP  ] 104c 8241 R:02 C:0c0330 (      usb-xhci) LOC_CODE=Backplane PLX
[   43.122200661,5] PHB#0001:02:03.0 [SWDN] 10b5 8718 R:aa C:060400 B:06..06 
[   43.122327813,5] PHB#0001:06:00.0 [EP  ] 14e4 1656 R:10 C:020000 (      ethernet) LOC_CODE=Backplane PLX
[   43.122471087,5] PHB#0001:06:00.1 [EP  ] 14e4 1656 R:10 C:020000 (      ethernet) LOC_CODE=Backplane PLX
[   43.122612671,5] PHB#0001:02:04.0 [SWDN] 10b5 8718 R:aa C:060400 B:07..07 
[   43.122722406,5] PHB#0001:07:00.0 [LGCY] 1b4b 9235 R:11 C:010601 (          sata) LOC_CODE=Backplane PLX
[   43.122867922,5] PHB#0002:00:00.0 [ROOT] 1014 03dc R:00 C:060400 B:01..ff SLOT=Slot1 
[   43.123092726,5] IPMI: Resetting boot count on successful boot
[   43.123217727,5] INIT: Waiting for kernel...
[   43.123297263,5] INIT: platform wait for kernel load failed
[   43.123375864,5] INIT: Assuming kernel at 0x20000000
[   43.123455071,3] INIT: ELF header not found. Assuming raw binary.
[   43.123544726,5] OCC: All Chip Rdy after 0 ms
[   43.125722586,4] NVRAM: Query before is done loading
[   43.125795263,4] NVRAM: Waiting for load
[   43.125832566,2] NVRAM: Failed to load
[   43.129735234,5] INIT: Starting kernel at 0x0, fdt at 0x305b6950 54517 bytes)
[   43.129890640,3] ***********************************************
[   43.129992054,3] Fatal Exception 0xe40 at 0000000000000000   
[   43.130108937,3] CFAR : 00000000300051c4
[   43.130191411,3] SRR0 : 00000000300028b0 SRR1 : 90000000001a1000
[   43.130306910,3] HSRR0: 0000000000000000 HSRR1: 9000000000001000
[   43.130447300,3] DSISR: 42000000         DAR  : 0000000001809000
[   43.130562885,3] LR   : 00000000300144d8 CTR  : 0000000000000000
[   43.130725951,3] CR   : 20002802         XER  : 00000000
[   43.130877503,3] GPR00: 0000000030014494 GPR16: 0000000000000000
[   43.130969202,3] GPR01: 0000000031ca3b80 GPR17: 0000000000000000
[   43.131108784,3] GPR02: 0000000030101600 GPR18: 0000000000000000
[   43.131200969,3] GPR03: 00000000305b6950 GPR19: 0000000000000000
[   43.131294257,3] GPR04: 0000000000000000 GPR20: 0000000000000000
[   43.131433394,3] GPR05: 0000000000000000 GPR21: 0000000000000000
[   43.131549178,3] GPR06: 0000000065504150 GPR22: 0000000000000000
[   43.131664841,3] GPR07: 0000000040000001 GPR23: 00000000300cb591
[   43.131804182,3] GPR08: 0000000030000000 GPR24: 00000000300b8362
[   43.131897015,3] GPR09: 00000000300050b0 GPR25: 0000000030301440
[   43.131988784,3] GPR10: 00000000000050b0 GPR26: 000000003031cc18
[   43.132105657,3] GPR11: 0000000000000000 GPR27: 0000000000000000
[   43.132220769,3] GPR12: 0000000040002808 GPR28: 00000000300b8201
[   43.132337603,3] GPR13: 0000000031ca0000 GPR29: 0000000040000000
[   43.132475925,3] GPR14: 0000000030002680 GPR30: 0000000030301600
[   43.132594081,3] GPR15: 0000000000000000 GPR31: 000000003031cbd8
[   43.132687590,0] Aborting!
CPU 0028 Backtrace:
 S: 0000000031ca3870 R: 0000000030013640   .backtrace+0x44
 S: 0000000031ca3910 R: 0000000030019e38   ._abort+0x4c
 S: 0000000031ca3990 R: 0000000030027b44   .exception_entry+0x18c
 S: 0000000031ca3b80 R: 0000000000001f10 * 
 S: 0000000031ca3d60 R: 0000000030014494   .load_and_boot_kernel+0xa30
 S: 0000000031ca3e30 R: 0000000030014bbc   .main_cpu_entry+0x614
 S: 0000000031ca3f00 R: 00000000300026e0   boot_entry+0x1b8
```

**Comment by mikey** [comment]
In that case we could just check that the 32 bits at 0x20000000 isn't all 0s and bail out.

Something like this could work (untested):
```
diff --git a/core/init.c b/core/init.c
index 5633a9d38f..a84c562c40 100644
--- a/core/init.c
+++ b/core/init.c
@@ -405,6 +405,10 @@ static bool load_kernel(void)
 
 	if (kh->ei_ident != ELF_IDENT) {
 		prerror("INIT: ELF header not found. Assuming raw binary.\n");
+
+		/* Check there is something there before we branch to it */
+		assert(*(uint32_t *)kh != 0);
+
 		return true;
 	}
 
```

**Comment by mikey** [comment]
How about this?
```
diff --git a/core/init.c b/core/init.c
index 5633a9d38f..c768dde9d1 100644
--- a/core/init.c
+++ b/core/init.c
@@ -405,6 +405,13 @@ static bool load_kernel(void)
 
 	if (kh->ei_ident != ELF_IDENT) {
 		prerror("INIT: ELF header not found. Assuming raw binary.\n");
+
+		/* Check there is something there before we branch to it */
+		if (*(uint32_t *)kh == 0) {
+			prerror("INIT: raw binary is zeros, can't execute!\n");
+			assert(0);
+		}
+
 		return true;
 	}
 
```
which gives this output:

```[    0.006650393,4] FLASH: Can't load resource id:0. No system flash found
[    0.006667359,4] FLASH: Can't load resource id:1. No system flash found
[    0.007001230,5] PCI: Resetting PHBs and training links...
[    0.007010769,5] PCI: Probing slots...
[    0.007183899,5] INIT: Waiting for kernel...
[    0.007188863,5] INIT: platform wait for kernel load failed
[    0.007194061,5] INIT: Assuming kernel at 0x20000000
[    0.007205006,3] INIT: ELF header not found. Assuming raw binary.
[    0.007210219,3] INIT: raw binary is zeros, can't execute!
[    0.007215088,0] Assert fail: core/init.c:412:0
[    0.007221311,0] Aborting!
CPU 0000 Backtrace:
 S: 0000000031c03ba0 R: 000000003001362c   .backtrace+0x48
 S: 0000000031c03c40 R: 0000000030019e64   ._abort+0x4c
 S: 0000000031c03cc0 R: 0000000030019ee0   .assert_fail+0x34
 S: 0000000031c03d40 R: 0000000030013c80   .load_and_boot_kernel+0x268
 S: 0000000031c03e30 R: 0000000030014bd0   .main_cpu_entry+0x610
 S: 0000000031c03f00 R: 00000000300026e0   boot_entry+0x1b8
```

**Comment by mikey** [comment]
Posted patch: http://patchwork.ozlabs.org/patch/879478/

**Comment by mikey** [comment]
Fixed in 18d7ee718bef3c95787473e3537be5f0653470c4
